// gatsby-plugin-webfonts/components/preload.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = Preload;

var _react = _interopRequireDefault(require("react"));

function Preload(_ref) {
  var css = _ref.css,
      disabled = _ref.disabled,
      _ref$format = _ref.format,
      format = _ref$format === void 0 ? "woff2" : _ref$format,
      _ref$crossOrigin = _ref.crossOrigin,
      crossOrigin = _ref$crossOrigin === void 0 ? "anonymous" : _ref$crossOrigin;
  if (disabled) return null;
  var regex = /url\((.+?)\)/gi;
  var fontUrls = css.match(regex).map(function (urlString) {
    return urlString.replace(regex, "$1");
  }).filter(function (urlString) {
    return urlString.endsWith("." + format);
  });
  return fontUrls.map(function (url, key) {
    return _react["default"].createElement("link", {
      key: "webfonts" + key,
      rel: "preload",
      as: "font",
      type: "font/" + format,
      crossOrigin: crossOrigin,
      href: url
    });
  });
}
